### ISSUE TITLE
"escapeDefault" in "godot.util.generator.util" causes errors

### ISSUE THREAD

**Issue by al1-ce** [godot-dlang/godot-dlang#29]
```src\godot\core\string.d(85,28): Error: union `godot.core.string.String._String` no size because of forward reference```

@Superbelko do you know how to fix this kind of thing? I'm pretty much stuck at this point since it's a D bug and debugging it is hell

**Comment by Superbelko** [comment]
https://github.com/godot-dlang/godot-dlang/blob/876e8ce9c66919fa904f511ba350441e36dd1ed2/src/godot/core/string.d#L80

make sure both godot_string and String_Bind is present and has definition. No other ideas from how it looks.

**Comment by al1-ce** [comment]
https://github.com/godot-dlang/godot-dlang/blob/876e8ce9c66919fa904f511ba350441e36dd1ed2/src/godot/core/string.d#L21 defines String_Bind
https://github.com/godot-dlang/godot-dlang/blob/876e8ce9c66919fa904f511ba350441e36dd1ed2/src/godot/core/string.d#L19 defines godot_string through `godot.c.core`

`godot_string` is struct with `size_t _opaque` and `String_Bind` is complex struct that's part of auto-generated bindings

**Comment by al1-ce** [comment]
Ok, ok, ok
I found out what causes that that wasn't there
https://github.com/godot-dlang/godot-dlang/blob/9db0554271f63cbaa14b5d7eeaae22316ff31f70/godot/util/generator/methods.d#L103-L106
By some crazy reason #7 breaks entire project

Love when fixing bugs creates more bugs

**Comment by al1-ce** [comment]
Honesly it's another point for having D strings converted inside of functions

**Comment by Superbelko** [comment]
As I said before, it was like this because of nogc-nothrow everywhere, sure you can do whatever you want, but there is always religious people who won't tolerate a GC and will come complain how bad it is and OOP sucks and so on.

**Comment by al1-ce** [comment]
I'm not rushing, just looking which parts of generetor would cause problems now

**Comment by al1-ce** [comment]
That actually might be a good solution for now

**Comment by al1-ce** [comment]
No, it seems generator needs to be rewritten at least partially. Coz having this kind of thing:
```d
struct String {
	string str;
	this(string s) {
		str = s;
	}
}
```
Now throws `` cannot implicitly convert expression `" "` of type `string` to `const(String)` ``
And trying to wrap it as `String(" ")` throws forward reference again

**Comment by Superbelko** [comment]
try `this(in string s)`

**Comment by al1-ce** [comment]
Did nothing. Tried doing thing I was talking about earlier:
```d
String lpad(in long min_length, in/* --> */string/* <-- */character = " ") const {
    ...
    return callBuiltinMethod!(String)(..., cast()/* --> */gs!/* <-- */character);
}
```
And it worked. I need to trace steps in `generator.methods` and find a way to get unescaped type and pass it into `body_`
It seems I can do something through `GodotArgument` struct
Complicated

**Comment by al1-ce** [comment]
Did litle hack, it works now, but, again, bunch of other types are screaming
There's going to be a lot of exceptions and conversions

**Comment by Superbelko** [comment]
Sorry, I misread this part 

```cannot implicitly convert expression `" "` of type `string` to `const(String)```

it actually needs const ctor right?

`this(in string s) const`

**Comment by al1-ce** [comment]
Const actually doesn't matter. I removed `in` in `in String character` and it still was saying no

It works now so don't bother about it
Though I'm hitting lots of other type errors now, but they're mild to fix

**Comment by al1-ce** [comment]
In godot-d this function is described as:
```d
Array buildCylinderPlanes(in double radius, in double height, in long sides, in long axis = 2)
```
Which is basically, I guess, the reason why it's like that

**Comment by al1-ce** [comment]
Ok, I see why templates
Compile-time lookup. Smart, forgot about it
Basically now compiler complains about being unable to read variables at compile time
Bandaided it with usual function

Also now there's errors about being unable to pass `String` into `string`
Is there even a reason aside `@nogs` to use `String` struct instead of default string?

**Comment by Superbelko** [comment]
> Also now there's errors about being unable to pass `String` into `string` Is there even a reason aside `@nogs` to use `String` struct instead of default string?

No I don't think so, but not 100% sure.

String struct itself is required, it can't be just dropped and forgotten because it is a part of core godot types, but I think it can have helper wrappers that takes plain D string arguments instead.
D `string` itself has nothing to do with GC, it was that gs! template to avoid runtime GC allocation.

However by adding these helpers it can quickly run out of hands if applied per argument basis like, so it is basically all or nothing situation. 

And of course it could be same issue with arrays as well.

**Comment by al1-ce** [comment]
Logically we just need fast GDTypeToDType and DTypeToGDType functions
The problem lies in the word "fast"

And I don't see where it'd spiral out of hand if it'd be something like that
```d
ptrcall!(void)(GDNativeClassBinding.method_setText, _godot_object, cast() dstringToGodotString(text));
```
For arrays I had some error earlier but adding T[] constructor solved that issue (not like string)

**Comment by al1-ce** [comment]
> As I said before, it was like this because of nogc-nothrow everywhere, sure you can do whatever you want, but there is always religious people who won't tolerate a GC and will come complain how bad it is and OOP sucks and so on.
> 
> **edit**: It is possible to have implicit construction in some places, unfortunately function parameter is not one of them... Just don't rush to convert all the generator.
> 
> ```d
> struct String
> {
> 	string str;
> 	
> 	
> 	this(string s)
> 	{
> 		str = s;
> 	}
> }
> 
> void print(String s = "default arg")
> {
> 	import std.stdio;
> 	writeln(s.str);
> }
> ```

I feel dumb. This didn't work because there was no `opAssign(in string)` overload

**Comment by al1-ce** [comment]
I swear to god D compiler demanding crazy paths to modules going to drive me insane

**Comment by al1-ce** [comment]
@Superbelko take a look (this is pure mess) [string-autoconversion](https://github.com/godot-dlang/godot-dlang/tree/string-autoconversion)

**Comment by al1-ce** [comment]
There's two ways how it can go
A: remove this branch and screw auto-converting ( I'd need to make a small fix in main to allow opAssing(string) )
B: merge this branch and try to make it as least impactful as possible and tidy it up overall

Btw those strange file moves are because of module names and D module searching

**Comment by Superbelko** [comment]
so basically auto conversion is only driven by that godotToD-whatever function? I quickly looked through patch and it looks okayish, if it allows plain string usage then why not.

also https://github.com/godot-dlang/godot-dlang/blob/string-autoconversion/src/godot/string.d#L327
from what I understand internally String and StringName is a same type so it might be possible to do cast to do cast and String.length() instead of that. 
But even if it it's not StringName is not supposed to be used as a regular string so probably ok. 
Also IIRC there is third option, in godot-cpp it instead takes StringName.data() (or was it buf()?) which is a simple CharString helper when strlen() can be used to get length.

In this situation I will probably just look at how C# did strings, if they are using .NET strings in public functions then we can go with variant B, otherwise prefer A.


Asteroids crash: run with debugger, probably it's the bindings now doing something crazy.

**Comment by al1-ce** [comment]
Looked up mono bindings
Thats what it has to say about StringName:
```c#
/// StringNames are immutable strings designed for general-purpose representation of unique names. 
/// StringName ensures that only one instance of a given name exists (so two StringNames with the 
/// same value are the same object). 
/// Comparing them is much faster than with regular strings, because only the pointers are compared, 
/// not the whole strings.
```

And internally it uses .net strings pretty much everywhere with addition of [StringExtensions.cs](https://github.com/godotengine/godot/blob/master/modules/mono/glue/GodotSharp/GodotSharp/Core/StringExtensions.cs)

So we probably better off with B option
It's going to be hell to make it pretty

**Comment by al1-ce** [comment]
Another thing with A option is that it was made for `@nogc` and it's using templates for everything
That ensures that almost everything is known at compile time, but when you are going to make big projects like that it's gonna look ugly
Basically template on a template on a template
That's why I want to make it work with D strings and classes for Godot types instead of structs
Yea, it looses a lot of safety and some speed, but we're talking non-specific (ak not handwritten for something) game engine that uses non-compiled language in it. So I think that my desires are somewhat justified

**Comment by SpiceyWolf** [comment]
cant you just do something like this?
```
struct String
{
	string str;
	
	this(string s)
	{
		str = s;
	}
	
	ref String opAssign(string value) return {
		this.str = value;
		return this;
	}
}
```

**Comment by al1-ce** [comment]
Already did in string-autoconversion branch

https://github.com/godot-dlang/godot-dlang/blob/string-autoconversion/src/godot/string.d#L100-L102

https://github.com/godot-dlang/godot-dlang/blob/string-autoconversion/src/godot/string.d#L153-L159

**Comment by al1-ce** [comment]
Fixed in 0626d5b
